flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -529,15 +542,19 @@ public void close() throws IOException {
     }
 
     public void cancel() throws IOException {
-        List<AsyncCheckpointRunnable> asyncCheckpointRunnables = null;
         synchronized (lock) {
             if (!closed) {
                 closed = true;
-                asyncCheckpointRunnables = new ArrayList<>(checkpoints.values());
+                final List<AsyncCheckpointRunnable> asyncCheckpointRunnables =
+                        new ArrayList<>(checkpoints.values());
                 checkpoints.clear();
+                if (!asyncCheckpointRunnables.isEmpty()) {
+                    asyncDisposeThreadPool.execute(
+                            () -> IOUtils.closeAllQuietly(asyncCheckpointRunnables));
+                }
+                asyncDisposeThreadPool.shutdown();
```

Comment:
What's the point of this construct? If clean up is taking long time, I don't think we should be hiding this fact when shutting down subtask, otherwise we run a risk of a resource leak. If subtask is not able to shutdown within a certain timeout, TM shut be killed and restarted (this is responsibility of the `TaskCancelerWatchDog`).

Reply:
Yes, you are right, I originally just hoped that all async operations should not block the process, but from the perspective of preventing resource leaks, cancel should not be asynchronous. I'll change this part of the logic back.